### DISCUSSION TITLE
Need help please: Building tests that @override_settings() to test new models

### DISCUSSION THREAD

**Discussion by ocuna** [wagtail/wagtail#8169] [General]
I'm trying to build my first patch to wagtail core.  I've created several tests but I need to figure out how to use @override_settings() to apply new settings and create a new model that can be used within the test.  After the patch is applied, a user would apply the setting and then use "makemigrations" with "migrate" to achieve in the database what I'm trying to test.

Please let me know if there is there any documentation on this that I could read, or code within tests I should be emulating.

Do I need to make a custom class here: 
`from wagtail.test.testapp.models import CustomImage, CustomImageFilePath`

Similar to how wagtail.images.tests.test.TestGetImageModel() works?

Thanks!

About Me: Django/Wagtail is amazing . . . Trying to migrate my Drupal 7 Enterprise site instead of going to Drupal 9

**Comment by lb-** [comment]
If you need to make new test models, here is some info on how to run the migrations

https://docs.wagtail.org/en/stable/contributing/developing.html#running-migrations-for-the-test-app-models

Here are the Django docs on overriding settings for a test

https://docs.djangoproject.com/en/4.0/topics/testing/tools/#django.test.override_settings

However, I'm not sure what your goals of the patch are but if you are proposing a new setting to be added I suggest that you create an issue first that explains the usecase. 

In general we are trying to avoid adding more settings and find other ways to achieve the goals of customisation where possible.

**Reply by ocuna** [reply]
Thanks for your direction.  I did eventually write a patch removing a dependency to change override settings, but when I tested it on a vanilla install, the problem the patch was intended for didn't occur.  Turns out:
- Original file names of renditions can be severely truncated depending on length of filter-spec attributes
- Duplicate file names of renditions become common under long filter-specs
- (patch solution) Increasing the length of file names reduces this possibility
- (actual solution) Setting overwrite_files in storages.storages.backends.azure_storage from True to False

Sheesh, about 3 days of wasted time - Learned a lot about testing though :-J

**Reply by lb-** [reply]
Hey,

Saw your other show and tell and just wanted to say thanks for sharing your learnings.

I find writing unit tests take me about triple the time (for something new) compared to actually writing a feature/fix but it usually really helps me find out more about the problem, even refine my solution and the best thing is it helps to ensure that the code does not break in the future.

Good on you for working out your specific solution also.

As per other comment, I think this journey of yours would make a great blog post, you also may help others who are using azure storage in the future also.

I use https://dev.to/ to blog as it is easy and public, plus I can write the blog in markdown and not have to get myself distracted by finding the 'perfect' static blog building tool haha.

Have a great rest of the week.

**Reply by ocuna** [reply]
Thanks for the encouragement!
